### PR TITLE
Database Output Report

### LINKED ISSUE #98: Add manure outputs to summary database
- [x] plan out new columns/tables
- [x] add connection to write manure outputs to the DB during the run
- [x] add variables to descriptions table
- [x] add options to export/manage data through web interface

### LINKED ISSUE #99: Add feed outputs to summary database
- [x] plan out new columns/tables
 - [x] add connection to write feed outputs to the DB during the run
 - [x] add variables to descriptions table
 - [x] add options to export/manage data through web interface

### LINKED ISSUE #100: Add Farm ES output formatting
- [x] plan out new columns/tables
- [x] add connection to write manure outputs to the DB during the run
- [x] add variables to descriptions table
- [x] add options to export/manage data through web interface (add summary to all years in report)

### PR DESCRIPTION
**Pull request by militsasotirova**
This pull request contains the output summary database. I created a report (DBReport) that works on the same principles as the other output reports except data is written to the `past_outputs.sqlite` database rather than a csv. Some methods from BaseReport are overwritten to avoid writing to a csv. 

In addition to the report name, you can also specify the version of the program (currently “pre_v1”) as well as the user’s name (currently “default”) in the JSON files. The idea is that we can track how the model evolves with development in terms of inputs/outputs. **Currently, the DBReport is turned off for all report JSON files.**

There is a folder `MASM/output_db` which holds `past_outputs.sqlite` as well as the code for the server that hosts the webpage through which you can interact with the database (run `python db_view.py` and follow the instructions printed to the console).  There’s more detailed information on functionality in the documentation & website itself, but essentially you can do 3 things through the webpage/UI: 

- export one or more result sets to a group of csv files (you can also select which columns you want to export), the input json file which corresponds to the input for the run that produced that result set(s), as well as a csv file that has more detailed descriptions of the columns that are stored in the database. All of these files are put in a folder under `MASM/output/db_output/[id_title]`.

- delete a result set from all tables (“permanent action”, unless you discard the changes made to past_outputs.sqlite locally)

- rename/re-title a result set 

Note that the actions become available once the appropriate amount of results are selected. There’s also an option to filter the results displayed. Here is the video I created last year that has a tutorial on the UI: 
https://3.basecamp.com/3486446/buckets/5296287/uploads/2939057141
It’s slightly outdated as I’ve added more variables since then but the general idea is the same.

There are currently some result sets in the database (most of which just contain 0s for the data), mainly for testing purposes. Feel free to play around with those using the UI. (Also, ideally the titles of the result sets would be unique and/or more descriptive - ‘rufas_test’ was just for testing purposes.) I’ll likely clear the data before I merge, this is here just for review.

The very first commit is actually the first commits I made for this code squashed, which is why it is so large. My original plan was to release that as a separate PR and then the additional changes as a second PR. However, in the code for the second wave I did, I cleaned up some of the code from the first so I figured it would be best to release both in one. Apologies for the massive size of the PR - let me know if you have any questions.

### CODE REVIEW

**Comment by militsasotirova** [comment]
Question for @PooyaHekmati - I see that there are 14 issues with CodeFactor, but when I click on "details" I get a Page Not Found error. How can I find out what the problems are?

**Comment by PooyaHekmati** [comment]
> Question for @PooyaHekmati - I see that there are 14 issues with CodeFactor, but when I click on "details" I get a Page Not Found error. How can I find out what the problems are?

It works for me. To unblock you, I am sending what I can see to you. We will figure out what is wrong later.

**Comment by militsasotirova** [comment]
I looked at the issues on CodeFactor, and most of them have to do with possible SQL injections, which are security issues (I thought [this article](https://realpython.com/prevent-python-sql-injection/) explained it well). [This part of the article](https://realpython.com/prevent-python-sql-injection/#using-sql-composition) provides a nice solution to some of the problems, however the Python library I'm using (sqlite3) does not provide such a way to safely compose SQL queries - at least, not that I could find. I'm thinking that these might not actually be an issue with the way I have structured the user inputs - the string snippets that I concatenate are totally controlled by the code. The user is only able to check boxes for the columns they want in the final output. However, I am not 100% confident saying that because I have never coded something like this in the past. (What kinds of attacks should we be wary of, if any?)

My other thought is that the database is stored locally, so this isn't an issue because if something goes wrong we can just restore a previous version of the data. 

I'm not quite sure where to go with this - any thoughts would be appreciated!

**Comment by militsasotirova** [comment]
> I looked at the issues on CodeFactor, and most of them have to do with possible SQL injections, which are security issues (I thought [this article](https://realpython.com/prevent-python-sql-injection/) explained it well). [This part of the article](https://realpython.com/prevent-python-sql-injection/#using-sql-composition) provides a nice solution to some of the problems, however the Python library I'm using (sqlite3) does not provide such a way to safely compose SQL queries - at least, not that I could find. I'm thinking that these might not actually be an issue with the way I have structured the user inputs - the string snippets that I concatenate are totally controlled by the code. The user is only able to check boxes for the columns they want in the final output. However, I am not 100% confident saying that because I have never coded something like this in the past. (What kinds of attacks should we be wary of, if any?)
> 
> My other thought is that the database is stored locally, so this isn't an issue because if something goes wrong we can just restore a previous version of the data.
> 
> I'm not quite sure where to go with this - any thoughts would be appreciated!

Update - we won't be worrying about this.

**Comment by militsasotirova** [comment]
I've finished adding what test cases I was able to - within the test files themselves I have provided explanations for why I was unable to test some of the functions. Any input on how I can get around my issues would be greatly appreciated!

**Comment by PooyaHekmati** [comment]
> 

How many issues do you have? Is it possible to collect them in one place?
There are two general types of issues with unit testing: (1) the unit under test is complex (it is more than one unit under the hood). The solution to this is to break the unit into smaller reasonable units. (2) The unit under test is coupled to the rest of the system. The solution to this is to remember that at this stage we are working on "unit" tests rather than integration tests. So, the units must be tested decoupled from the rest of the system. This can be achieved by pathing and mocking the unit's interactions with the rest of the system. Maybe this helps? https://changhsinlee.com/pytest-mock/

**Comment by militsasotirova** [comment]
After looking at the link you sent, I'm not sure that mocking will help for the tests in `test_db_report.py` (there's more detailed comments within the file itself as to why I'm not sure how to test them - mostly has to do with when I actually commit changes to the database during a simulation execution). For the problem I'm facing in `test_db_output.py` - I might be misunderstanding, but mocking seems to be for replacing the contents within a function or a class so that testing is more efficient in some way. In my case, I am not able to instantiate a `RequestHandler` in order to even call the appropriate methods in tests (also more details in the comments of the file).

File: tests/test_db_report.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    """Unit test for function initialize in file
+    RUFAS/output_handler/reports/db_report.py"""
+    # Nothing to test as this function explicitly does nothing.
+    assert True
+
+
+def test_write_headers():
+    """Unit test for function write_headers in file
+    RUFAS/output_handler/reports/db_report.py"""
+    # Nothing to test as this function explicitly does nothing.
```
Comment (PooyaHekmati) [inline tests/test_db_report.py]: For the function whose job is to ensure a similarly named function is not called, you need to check that this function does its job by something like assert_not_called or assert.call_counts == 0

File: tests/test_db_report.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+def test_finalize():
+    """Unit test for function finalize in file
+    RUFAS/output_handler/reports/db_report.py"""
+    # I wasn't sure how to test this.
+    pass
+
+
+def test_store_results_setup():
+    """Unit test for function store_results_setup in file
+    RUFAS/output_handler/reports/db_report.py"""
```
Comment (PooyaHekmati) [inline tests/test_db_report.py:120]: You need to mock the connection and db.
https://stackoverflow.com/questions/18077247/how-to-mock-sqlite3-connect-in-python

File: tests/test_db_report.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+def test_produce_report_graphics():
+    """Unit test for function initialize in file
+    RUFAS/output_handler/reports/db_report.py"""
+    # Nothing to test as this function explicitly does nothing.
+    assert True
+
+
+def test_finalize():
+    """Unit test for function finalize in file
```
Comment (PooyaHekmati) [inline tests/test_db_report.py:101]: by checking how many times self.conn.commit() has been called. This can be done using mocking.

**Comment by PooyaHekmati** [comment]
> After looking at the link you sent, I'm not sure that mocking will help for the tests in `test_db_report.py` (there's more detailed comments within the file itself as to why I'm not sure how to test them - mostly has to do with when I actually commit changes to the database during a simulation execution). For the problem I'm facing in `test_db_output.py` - I might be misunderstanding, but mocking seems to be for replacing the contents within a function or a class so that testing is more efficient in some way. In my case, I am not able to instantiate a `RequestHandler` in order to even call the appropriate methods in tests (also more details in the comments of the file).

Mocking is not for efficiency, we mock external dependencies and connections so that we can isolate the unit and test it. A unit test should not have any side effects; i.e., the state of the system should be the same before and after the test. It is NOT to really create a file or make an HTTP call or create a db etc. All of this should be simulated using mocking so from the perspective of the unit all of these are happening but it is just a simulation.

**Comment by PooyaHekmati** [comment]
Closing the PR as it has become stale and we have changed directions since this PR was sent out.

**Pull request closed without merging**